Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py

```diff
@@ -0,0 +1,379 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+#   NScD Oak Ridge National Laboratory, European Spallation Source,
+#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+# SPDX - License - Identifier: GPL - 3.0 +
+
+from mantid import config
+from mantid.api import AlgorithmFactory, FileAction, FileProperty, \
+    ITableWorkspaceProperty, Progress, PythonAlgorithm
+from mantid.kernel import Direction, IntArrayBoundedValidator, \
+    StringListValidator, StringMandatoryValidator
+from mantid.simpleapi import *
+
+import fnmatch
+import h5py
+import numpy as np
+import os
+import re
+
+
+class GenerateLogbook(PythonAlgorithm):
+
+    _data_directory = None
+    _facility = None
+    _instrument = None
+    _numor_range = None
+    _metadata_headers = None
+    _metadata_entries = None
+
+    def category(self):
+        return 'Utility'
+
+    def summary(self):
+        return 'Generates logbook containing meta-data specific to the instrument and technique used to obtain the raw data'
+
+    def name(self):
+        return 'GenerateLogbook'
+
+    def validateInputs(self):
+        issues = dict()
+
+        instrument = self.getPropertyValue('Instrument')
+        ws_tmp = CreateSingleValuedWorkspace()
+        try:
+            LoadParameterFile(Workspace=ws_tmp, Filename=instrument + '_Parameters.xml')
+        except Exception as e:
+            self.log().error(str(e))
+            issues['Instrument'] = 'There is no parameter file for {} instrument.'.format(instrument)
+        DeleteWorkspace(Workspace=ws_tmp)
+
+        if not self.getProperty('NumorRange').isDefault:
+            numor_range = self.getProperty('NumorRange').value
+            if len(numor_range) < 2:
```

Comment:
Could you validate that the length is equal to 2 here? Unless there are some use cases I've not thought about where you accept multiple ranges. Seems like you just read the [0] and [-1] elements though

Reply:
Unfortunately, in that case all range specifiers similar to those used in `MultiFileProperty` would not be allowed. The implemented solution is not the most elegant, as it creates a list based on the input. If two elements are provided, that naturally gives the limits for the numor range. If a range is specified, e.g. through "360:365" or equally "360-365", a list of `int` containing all values ranging from 360 to 365, inclusive, is created and the code takes the first and last values as the numor range. Such a list is going to be longer than the desired 2 element, and during validation, the property will return a list, not the user input, so I am not sure here how to validate against it. A solution could be to provide more comprehensive tooltip.

Reply:
If the ranges are being expanded out for you by the C++ code then I think you can just use the list in _prepare_file_array rather than extracting the [0] and [-1] elements ie
```
            try:
                numor = int(os.path.splitext(file[instrument_name_len:])[0])
                if numor in self._numor_range:
                    file_list.append(os.path.splitext(file)[0])
            except (ValueError, OverflowError):
                self.log().debug("File {} cannot be cast into an integer numor".format(file))
                continue
```
I just tried this and seemed to work. I saw a problem with filenames containing strings like 394458_394882 which seems to get successfully converted to an integer so that's why I added the OverflowError handling.

I think this would mean the "," no longer indicates a range and instead just indicates a list. But that seems OK and consistent with the "," elsewhere in Mantid?

Reply:
Thank you for this suggestion, this is pretty neat. In this form, a list with all desired numors can be provided and the file names are going to be verified against it. The user can also provide a range of numors that will be turned into a list which can be used to validate against. A slight modification of the behaviour when the `NumorRange` property is not specified is required, but that's all. Now, this modification can support more robust workflow.